### DISCUSSION TITLE
CDN Caching recommendations

### DISCUSSION THREAD

**Discussion by EmmanuelPonnudurai** [graphql/graphql-js#4230] [Q&A]
What is the official recommendation from the graphql org team on CDN caching? 

With Graphql inherently being a HTTP POST, it makes it rather hard to set up CDN caching rules at edge servers. There are products which help achieve this. Some of them I have listed below,

- https://stellate.co/
- https://wundergraph.com/blog/open_source_graphql_cdn_edge_cache_with_cloudflare_fastly_fly_io

Coming to actual edge servers, some of them like Fastly, Akamai, cloudflare have varying levels of support to allow caching of Graphql HTTP post. But it also means vendor lock in and in some cases a complete re write in the way things are managed from the application standpoint. 

I am looking for some guidance in terms of CDN caching in general. The docs do not seem to list anything about this (or may be I didn't look properly). The only thing I saw is https://graphql.org/learn/caching/ and it is geared towards the client caching but not CDN caching. 

Since CDN caching and HTTP mechanisms to allow caching are universally used by several parties involved, often times with a rule of only HTTP GET, I feel like this is an area which is a bit obscure when it comes to GraphQL and looking for any valuable information which helps achieve this goal without vendor lock ins or large changes to existing system.

OR, is it safe to assume that CDN caching is out of scope for graphql org and each implementor should do as they see fit to accommodate it? 

Any recommendations and guidance will be much appreciated from the community in this regard. Thank you.

**Comment by n1ru4l** [comment]
How some servers do caching:
- Yoga https://the-guild.dev/graphql/yoga-server/docs/features/response-caching
- Apollo Server https://www.apollographql.com/docs/apollo-server/performance/caching

SaaS Solutions:
- Stellate https://stellate.co/graphql-edge-caching

**Comment by SX110903** [comment]
lo mas practico sin pagar por servicios externos es usar GET para queries en vez de POST, graphql lo soporta aunque casi nadie lo usa:

```
GET /graphql?query={users{id name}}&operationName=GetUsers
```

con GET puedes usar cache-control normal y cualquier CDN lo cachea sin problema. el limite es que las queries largas pueden no caber en la url pero para la mayoria de casos va bien

si necesitas algo mas serio persisted queries es el camino, registras las queries con un hash y el cliente manda solo el hash, asi puedes hacer GET con el hash y cachear a nivel de CDN sin exponer la query. apollo tiene soporte nativo para esto y la mayoria de clientes tambien

**Comment by yudin-s** [comment]
GraphQL can work with CDN caching, but I would not try to cache arbitrary POST bodies.

The usual production pattern is:

1. Use persisted queries for cacheable reads.
2. Send those persisted queries as GET requests.
3. Keep mutations and user-specific reads on POST / no-cache.

Your cache key should include at least:

- operation id / persisted query hash
- normalized variables
- auth scope or tenant/public/private scope
- locale/currency if relevant

For example:

```text
GET /graphql?operation=ProductList&hash=abc123&vars=...
Cache-Key: ProductList:abc123:varsHash:public:en-US
```

The main risk is accidentally caching personalized data. I would start with anonymous/public queries only, short TTLs, and explicit `Cache-Control` from the resolver/application layer. Then expand carefully to semi-private scopes where the auth scope is part of the key.

Also make sure errors are not cached the same way as successful responses. A temporary upstream error cached at the CDN can be very confusing with GraphQL because the HTTP status is often still 200.
